components/style/data.rs

```diff
@@ -8,24 +8,191 @@ use properties::ComputedValues;
 use selector_impl::PseudoElement;
 use std::collections::HashMap;
 use std::hash::BuildHasherDefault;
+use std::mem;
+use std::ops::{Deref, DerefMut};
 use std::sync::Arc;
 
-pub type PseudoStyles = HashMap<PseudoElement, Arc<ComputedValues>,
-                                BuildHasherDefault<::fnv::FnvHasher>>;
-pub struct PersistentStyleData {
+type PseudoStylesInner = HashMap<PseudoElement, Arc<ComputedValues>,
+                                 BuildHasherDefault<::fnv::FnvHasher>>;
+#[derive(Clone, Debug)]
+pub struct PseudoStyles(PseudoStylesInner);
+
+impl PseudoStyles {
+    pub fn empty() -> Self {
+        PseudoStyles(HashMap::with_hasher(Default::default()))
+    }
+}
+
+impl Deref for PseudoStyles {
+    type Target = PseudoStylesInner;
+    fn deref(&self) -> &Self::Target { &self.0 }
+}
+
+impl DerefMut for PseudoStyles {
+    fn deref_mut(&mut self) -> &mut Self::Target { &mut self.0 }
+}
+
+/// The styles associated with a node, including the styles for any
+/// pseudo-elements.
+#[derive(Clone, Debug)]
+pub struct NodeStyles {
     /// The results of CSS styling for this node.
-    pub style: Option<Arc<ComputedValues>>,
+    pub primary: Arc<ComputedValues>,
 
     /// The results of CSS styling for each pseudo-element (if any).
-    pub per_pseudo: PseudoStyles,
+    pub pseudos: PseudoStyles,
 }
 
-impl PersistentStyleData {
-    pub fn new() -> Self {
-        PersistentStyleData {
-            style: None,
-            per_pseudo: HashMap::with_hasher(Default::default()),
+impl NodeStyles {
+    pub fn new(primary: Arc<ComputedValues>) -> Self {
+        NodeStyles {
+            primary: primary,
+            pseudos: PseudoStyles::empty(),
+        }
+    }
+}
+
+#[derive(Debug)]
+enum NodeDataStyles {
+    /// The field has not been initialized.
+    Uninitialized,
+
+    /// The field holds the previous style of the node. We represent it this
+    /// way, rather than Previous(Option<NodeStyles>), to improve packing.
+    ///
+    /// This is the input to the styling algorithm. It would ideally be
+    /// immutable, but for now we need to mutate it a bit before styling to
+    /// handle animations.
+    EmptyPrevious,
+    Previous(NodeStyles),
```

Comment:
I think `size_of::<NodeStyles>()` should be equal to `size_of::<Option<NodeStyles>>`, since it contains an arc, and the discriminant of the option becomes the nullness of the `Arc`: https://play.rust-lang.org/?gist=bc261480736b63367ed39cc9c4feee82&version=stable&backtrace=0

So I think if it's clearer we should just use `Option<NodeStyles>`. Otherwise remove the comment about packing.